sdk/plugin/plugin.go

```diff
@@ -69,36 +77,72 @@ func NewBackend(ctx context.Context, pluginName string, pluginType consts.Plugin
 	return backend, nil
 }
 
-func NewPluginClient(ctx context.Context, sys pluginutil.RunnerUtil, pluginRunner *pluginutil.PluginRunner, logger log.Logger, isMetadataMode bool) (logical.Backend, error) {
-	// pluginMap is the map of plugins we can dispense.
-	pluginSet := map[int]plugin.PluginSet{
+// pluginSet returns the go-plugin PluginSet that we can dispense. This ensures
+// that plugins that don't support AutoMTLS are run on the appropriate version.
+func pluginSet(autoMTLS, metadataMode bool) map[int]plugin.PluginSet {
+	if autoMTLS {
+		return map[int]plugin.PluginSet{
+			5: {
+				"backend": &GRPCBackendPlugin{
+					MetadataMode:      false,
+					AutoMTLSSupported: true,
+				},
+			},
+		}
+	}
+	return map[int]plugin.PluginSet{
 		// Version 3 used to supports both protocols. We want to keep it around
 		// since it's possible old plugins built against this version will still
 		// work with gRPC. There is currently no difference between version 3
 		// and version 4.
 		3: {
 			"backend": &GRPCBackendPlugin{
-				MetadataMode: isMetadataMode,
+				MetadataMode: metadataMode,
 			},
 		},
 		4: {
 			"backend": &GRPCBackendPlugin{
-				MetadataMode: isMetadataMode,
+				MetadataMode: metadataMode,
 			},
 		},
 	}
+}
 
-	namedLogger := logger.Named(pluginRunner.Name)
-
+func NewPluginClient(ctx context.Context, pluginRunner *pluginutil.PluginRunner, config pluginutil.PluginClientConfig) (logical.Backend, error) {
 	var client *plugin.Client
 	var err error
-	if isMetadataMode {
-		client, err = pluginRunner.RunMetadataMode(ctx, sys, pluginSet, handshakeConfig, []string{}, namedLogger)
-	} else {
-		client, err = pluginRunner.Run(ctx, sys, pluginSet, handshakeConfig, []string{}, namedLogger)
+	merr := &multierror.Error{}
+	ps := pluginSet(config.AutoMTLS, config.IsMetadataMode)
+
+	// Best effort attempt to run with the version 5 plugin set which supports
+	// autoMTLS if that's the set returned
+	client, err = pluginRunner.RunConfig(ctx,
+		pluginutil.Runner(config.Wrapper),
+		pluginutil.PluginSets(ps),
+		pluginutil.HandshakeConfig(handshakeConfig),
+		pluginutil.Env(),
+		pluginutil.Logger(config.Logger),
+		pluginutil.MetadataMode(config.IsMetadataMode && !config.AutoMTLS),
+		pluginutil.AutoMTLS(config.AutoMTLS),
+	)
+	// This is the fallback: If we error with a v5 plugin set then try again,
+	// but force the old plugin set
+	if config.AutoMTLS && err != nil {
+		merr = multierror.Append(merr, err)
```

Comment:
I wonder if there is any value in logging (probably debug level) an error that occurred when trying to run the version 5 plugin set? Thinking it could be valuable to indicate why the version 5 `RunConfig()` failed and the fallback was successful. Feel free to disregard if you don't think there is value.

Reply:
This caused me to rethink this a bit more. This fallback doesn't add any value. This codepath is not being hit currently. It likely will not ever be hit, since all this is doing is returning a configured client. I think it is better to just configure this client how the caller wants and attempt to run it. An error will be returned if the configured client can't be connected to via rpc.